### ISSUE TITLE
Performance degradation

### ISSUE THREAD

**Issue by jbaron** [denizyuret/Knet.jl#562]
Noticed some serious performance degradation from Knet 1.3.0 to 1.3.5. I have developed a small framework to make it a bit easier to define and train DL networks and use Knet as the underlying engine (called Photon.jl). Part of this framework is a performance test using a small convolution network to test if my own code doesn't degrade too much. 

However if I keep my code the same and only change the version of Knet used, there is a large difference (17.7 seconds to 23.8 seconds) in performance. Any thoughts on this?

Here some of the details of the tests executed:

Knet version 1.3.0
 ================== 
  [1902f260] - Knet v1.3.0
  [79e6a3ab] - Adapt v1.0.1
  [6710c13c] - AutoGrad v1.2.2
  [fa961155] - CEnum v0.3.0
  [3895d2a7] - CUDAapi v1.2.0
  [c5f51814] - CUDAdrv v3.1.0
  [be33ccc6] - CUDAnative v2.4.0
  [3a865a2d] - CuArrays v1.2.1
  [1a297f60] - FillArrays v0.7.4
  [0c68f7d7] - GPUArrays v1.0.4
  [1902f260] - Knet v1.3.0
  [929cbde3] - LLVM v1.4.1
  [872c559c] - NNlib v0.6.6
  [a759f4b9] - TimerOutputs v0.5.5

julia --project=. performance/photon_conv.jl 
[   1:    100] - loss: 1.0347
[   2:    200] - loss: 1.0216
[   3:    300] - loss: 0.9665
[   4:    400] - loss: 0.8954
[   5:    500] - loss: 0.8222
[   6:    600] - loss: 0.7758
[   7:    700] - loss: 0.7431
[   8:    800] - loss: 0.7028
[   9:    900] - loss: 0.6669
[  10:   1000] - loss: 0.6829
[  11:   1100] - loss: 0.6478
 **17.691484 seconds** (5.39 M allocations: 271.628 MiB, 1.28% gc time)
 

 Knet version 1.3.5
 ==================
  [1902f260] + Knet v1.3.5
  [79e6a3ab] + Adapt v1.0.1
  [6710c13c] + AutoGrad v1.2.2
  [fa961155] + CEnum v0.3.0
  [3895d2a7] + CUDAapi v4.0.0
  [c5f51814] + CUDAdrv v6.3.0
  [be33ccc6] + CUDAnative v3.1.0
  [da1fd8a2] + CodeTracking v0.5.11
  [f68482b8] + Cthulhu v1.1.1
  [3a865a2d] + CuArrays v2.2.0
  [e2ba6199] + ExprTools v0.1.1
  [0c68f7d7] + GPUArrays v3.3.0
  [61eb1bfa] + GPUCompiler v0.2.0
  [1902f260] + Knet v1.3.5
  [929cbde3] + LLVM v1.4.1
  [872c559c] + NNlib v0.6.6
  [a759f4b9] + TimerOutputs v0.5.5

julia --project=. performance/photon_conv.jl 
[   1:    100] - loss: 1.0242
[   2:    200] - loss: 0.9408
[   3:    300] - loss: 0.9062
[   4:    400] - loss: 0.8609
[   5:    500] - loss: 0.7817
[   6:    600] - loss: 0.7587
[   7:    700] - loss: 0.6932
[   8:    800] - loss: 0.6314
[   9:    900] - loss: 0.6206
[  10:   1000] - loss: 0.5979
[  11:   1100] - loss: 0.5951
 **23.786814 seconds** (4.88 M allocations: 263.234 MiB, 0.89% gc time)

**Comment by denizyuret** [comment]
Knet uses CuArrays for array allocation, which may be responsible. To tell which package is responsible could you try `Knet.cuallocator()=false` and rerun the experiments? This reverts back to the old Knet allocator. If the performance degredation disappears, then something that happened in CuArrays may be responsible.

If not, there is a way to turn on timing for AutoGrad and Knet that may help pinpoint the issue:
1. set ENV["KNET_TIMER"] to "true" and rebuild Knet. After the experiment Knet.to should give us the timing info for GPU calls.
2. Set ENV["AUTOGRAD_TIMER"]="true" and Pkg.build("AutoGrad"). After the experiment AutoGrad.to should give us the timing per backprop operations.

**Comment by jbaron** [comment]
I added the following to the performance script

```
import Knet
Knet.cuallocator()=false

```

and the result was much worse for 1.3.5. So looks like there might also be an issue in the Knet memory allocator (seems to be a gc issue) in later versions. So I cannot really tell if the Cuarray mem allocator degradaded and is the root problem for my initial problem.

If I have some more time will see if I can do some more investigation. I think what would be really nice is to have a simple script that measures performance of the main building blocks of Knet and have it run with every release.

Knet version 1.3.5
==================
julia --project=. performance/photon_conv.jl 
[   1:    100] - loss: 0.8013
[   2:    200] - loss: 0.7748
[   3:    300] - loss: 0.7355
[   4:    400] - loss: 0.7182
[   5:    500] - loss: 0.7009
[   6:    600] - loss: 0.6344
[   7:    700] - loss: 0.5937
[   8:    800] - loss: 0.5394
[   9:    900] - loss: 0.5250
[  10:   1000] - loss: 0.5284
[  11:   1100] - loss: 0.5054
 **44.442761** seconds (4.40 M allocations: 240.039 MiB, 60.56% gc time)
 
 
 Knet version 1.3.0
===================
julia --project=. performance/photon_conv.jl 
[   1:    100] - loss: 0.9996
[   2:    200] - loss: 0.9747
[   3:    300] - loss: 0.9332
[   4:    400] - loss: 0.8933
[   5:    500] - loss: 0.8332
[   6:    600] - loss: 0.7768
[   7:    700] - loss: 0.7084
[   8:    800] - loss: 0.6794
[   9:    900] - loss: 0.6696
[  10:   1000] - loss: 0.6464
[  11:   1100] - loss: 0.6380
**18.614246** seconds (4.34 M allocations: 238.871 MiB, 4.51% gc time)

**Comment by jbaron** [comment]
Just a quick update, did some more testing, and it looks like cudnnConvolutionBackwardData  seems to be the biggest difference. Attached the logging details.

[photon_knet_logging.txt](https://github.com/denizyuret/Knet.jl/files/4674108/photon_knet_logging.txt)

**Comment by jbaron** [comment]
I did some more debugging by trying to find the minimal model that causes this behavior and I found out it only happens when i add two Conv layers:

  ```
model = Sequential(
    Conv2D(16, (3,3), relu),
    Conv2D(16, (3,3), relu),
  )
```

If I don't add the second conv layer (even adding other types of layers), the performance between 1.3.0 and 1.3.5 seems near identical. Under the hood I just invoke conv4 with the default cudnnhandle(). Any thoughts?

**Comment by jbaron** [comment]
Was able to reproduce performance degradation with regulat Knet app. Tested it with small script slightly changed from the home page. Performance was 6.8s on 1.3.0 and 8.4s on 1.3.5. 

I did notice that it really depends on the model architecture, so possible won't impact most models.

`
using Knet
using Statistics

struct Conv; w; b; f; end
(c::Conv)(x) = c.f.(conv4(c.w, x, padding = 0, stride = 1, dilation = 1) .+ c.b)
Conv(w1,w2,cx,cy,f=relu) = Conv(param(w1,w2,cx,cy), param0(1,1,cy,1), f)

struct Chain; layers; end
(c::Chain)(x) = (for l in c.layers; x = l(x); end; x)

(c::Chain)(x,y) = mean(c(x))

include(Knet.dir("data","mnist.jl"))
dtrn, dtst = mnistdata()

LeNet = Chain((Conv(3,3,1,16), Conv(3,3,16,16)))

sgd!(LeNet, repeat(dtrn,1))
@time sgd!(LeNet, repeat(dtrn,10))
`

**Comment by denizyuret** [comment]
I can confirm that the degradation is not due to changes in CuArrays, using your script:
```
Knet    CuArr   seconds
1.3.0   1.2.1   18.34, 18.39, 18.34, 18.29
1.3.5   2.2.0   21.74, 21.56, 21.70, 21.71
1.3.5   1.2.1   21.70, 21.66, 21.65, 21.74
```
There are two relevant changes from Knet-1.3.0 to Knet-1.3.5:
* Using `cudnnFindConvolutionBackwardDataAlgorithmEx` instead of `cudnnFindConvolutionBackwardDataAlgorithm` from cudnn library (same with forward). These functions pick the fastest algorithm for a given array size and the behavior of the two may be different.
* Using a single temporary `workSpace` in 1.3.0 to allocating a new `workSpace` for each cudnn convolution call. I was not sure how safe it is to use a single `workSpace` esp. when multiple streams may run in parallel on a single GPU. Also, having a common `workSpace` meant keeping the largest sized one which may waste space.

I will not investigate which one(s) are responsible for the degradation.

**Comment by jbaron** [comment]
Thanks for the help. Since it seems only applicable to very specific network architectures and possible due to underlying cudnn working, I think indeed not worth the trouble to further invest/optimize. So if OK with you, I'll close this one?

**Comment by denizyuret** [comment]
I think this effects all convolution operations, so it is worth spending some time on it. I also like your performance testing idea: "I think what would be really nice is to have a simple script that measures performance of the main building blocks of Knet and have it run with every release." -- I have a couple of old scripts under Knet/prof, but they are outdated and/or not used regularly. Let's keep it open for now.

**Comment by denizyuret** [comment]
Can you test with latest master?

**Comment by jbaron** [comment]
Sure, I'll get the Knet package from the master branch from github

**Comment by denizyuret** [comment]
`pkg"dev Knet"` should do it. Then you can go back using `pkg"free Knet"` I think.

**Comment by jbaron** [comment]
Thanks!!! I did a `add https://github.com/denizyuret/Knet.jl` which also works it seems (didn't even know about the `dev` option). 

And the  good news, we are back with very close to 1.3.0 performance :1st_place_medal:  And also the unit test in my package run all fine against the master branch (including some Densenet and VGG models).

Knet master 
=========================

julia --project=. performance/photon_conv.jl 
[   1:    100] - loss: 1.0150
[   2:    200] - loss: 1.0256
[   3:    300] - loss: 0.9571
[   4:    400] - loss: 0.9077
[   5:    500] - loss: 0.8395
[   6:    600] - loss: 0.8038
[   7:    700] - loss: 0.7450
[   8:    800] - loss: 0.7050
[   9:    900] - loss: 0.6783
[  10:   1000] - loss: 0.6474
[  11:   1100] - loss: 0.6193

 **18.259689** seconds (4.85 M allocations: 260.986 MiB, 1.17% gc time)

For your info, this is also close to PyTorch which of course does the autograd directly in C/C++ and has a much larger code base (and TF/Keras is typically substantial slower).

**Comment by denizyuret** [comment]
Great news! There are two tricks that did it: picking better algorithms, and using the same workSpace for all conv operations rather than reallocating every operation. I am not sure the second trick is safe (when running multiple streams etc), I asked some experts and may have to remove it if they say no. That adds about 1 second to the runtime (~5% degradation). Will decide before next release.

**Comment by jbaron** [comment]
Not too familiar with CUDA, but I assume workspace is used for reserving memory (and possible some other resources like threads) for CUDA operations. So the best would be pooling them (in case they are not multi-stream safe). 

But 5% degradation is also not too bad.  I think there more than enough room in my own Photon.jl code to gain more than 5%.
